src/screens/ParadeMapScreen/Map.js

```diff
@@ -0,0 +1,195 @@
+// @flow
+import React, { Component } from "react";
+import type { ElementRef } from "react";
+import {
+  Image,
+  Platform,
+  View,
+  StyleSheet,
+  TouchableWithoutFeedback,
+  Linking,
+  Alert
+} from "react-native";
+import MapView, { Polyline, Marker } from "react-native-maps";
+import Permissions from "react-native-permissions";
+import Text from "../../components/Text";
+import { velvetColor } from "../../constants/colors";
+import type {
+  Coordinates,
+  Region,
+  Terminals
+} from "../../constants/parade-coordinates";
+
+import locationButtonInactive from "../../../assets/images/location-inactive.png";
+import locationButtonActive from "../../../assets/images/location-active.png";
+
+type PermissionStatus = "authorized" | "denied" | "restricted" | "undetermined";
+const neverAskStatuses = new Set(["restricted"]);
+
+type Props = {
+  route: Array<Coordinates>,
+  paradeRegion: Region,
+  terminals: Array<Terminals>
+};
+
+type State = {
+  locationPermission: ?PermissionStatus,
+  atUserLocation: boolean
+};
+
+class Map extends Component<Props, State> {
+  state = {
+    locationPermission: undefined,
+    atUserLocation: false
+  };
+
+  componentDidMount() {
+    this.locationPermissionPromise = Permissions.check("location");
+    this.locationPermissionPromise.then(response => {
+      this.setState({ locationPermission: response });
+    });
+  }
+
+  onRegionChange = (position: Coordinates) => {
+    const { latitude: currentLat, longitude: currentLong } = position;
+    if (this.state.atUserLocation) {
+      this.setState({ atUserLocation: false });
+    } else if (this.state.locationPermission === "authorized") {
+      navigator.geolocation.getCurrentPosition(({ coords }) => {
+        const { latitude: userLat, longitude: userLong } = coords;
+        if (
+          userLat.toFixed(5) === currentLat.toFixed(5) &&
+          userLong.toFixed(5) === currentLong.toFixed(5)
+        ) {
+          this.setState({ atUserLocation: true });
+        }
+      });
+    }
+  };
+
+  locationPermissionPromise: Promise<PermissionStatus>;
+
+  confirmLocationPermission = async (): Promise<PermissionStatus> => {
```

Comment:
Part of this would add two more states to the `PermissionStatus` type, "checking" and "requesting". Adding these two states would give the UI the ability to disable buttons and feedback to the user when they can perform an action.